lib/fleek-ipld/src/walker/dag_pb.rs

```diff
@@ -0,0 +1,283 @@
+use std::ops::Deref;
+
+use async_trait::async_trait;
+use bytes::{Bytes, BytesMut};
+use ipld_core::cid::Cid;
+use ipld_core::ipld::Ipld;
+use ipld_dagpb::PbNode;
+use reqwest::Url;
+
+use super::processor::{DocId, IpldItem, IpldStream, Link, Processor};
+use crate::errors::IpldError;
+use crate::unixfs::Data;
+
+/// Processor for DAG-PB nodes in IPFS with UnixFS data
+#[derive(Clone)]
+pub struct IpldDagPbProcessor {
+    ipfs_url: Url,
+}
+
+struct PbNodeWrapper(PbNode);
+
+impl From<PbNode> for PbNodeWrapper {
+    fn from(node: PbNode) -> Self {
+        Self(node)
+    }
+}
+
+impl From<PbNodeWrapper> for Vec<Link> {
+    fn from(node: PbNodeWrapper) -> Self {
+        node.0
+            .links
+            .into_iter()
+            .map(|x| Link::new(x.cid, x.name, x.size))
+            .collect()
+    }
+}
+
+impl Deref for PbNodeWrapper {
+    type Target = PbNode;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl PbNodeWrapper {
+    pub fn from_bytes(bytes: &Bytes) -> Result<Self, IpldError> {
+        let node = PbNode::from_bytes(bytes.clone())?;
+        Ok(node.into())
+    }
+}
+
+impl IpldDagPbProcessor {
+    /// Creates a new `IpldDagPbProcessor` with the given IPFS URL.
+    ///
+    /// # Arguments
+    ///
+    /// * `ipfs_url` - A string slice that holds the URL of the IPFS node.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the provided IPFS URL is invalid.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// let processor = IpldDagPbProcessor::new("http://localhost:5001");
+    /// ```
+    pub fn new(ipfs_url: &str) -> Self {
+        Self {
+            ipfs_url: Url::parse(ipfs_url)
+                .unwrap_or_else(|_| panic!("Invalid IPFS URL {}", ipfs_url)),
+        }
+    }
+
+    async fn request(&self, cid: &Cid) -> Result<Bytes, IpldError> {
+        let url = self.ipfs_url.clone();
+        let url = url.join(&format!("ipfs/{}?format=raw", cid))?;
+        let response = reqwest::get(url).await?;
+        response.bytes().await.map_err(Into::into)
+    }
```

Comment:
It seems like we're making an individual request for each ipld block here, however it would be much faster to gather blocks from a streamed CAR file and lazily evaluate whole files/folders from them as we get each block. Since car archives would include every ipld block needed for the root cid, and we'd no longer have to deal with TLS roundtripping in between each small piece of data, we'd maintain better network throughput and reduce latency on cold content

Reply:
Nice tip. I will address this in the next PR where I will also improve memory management.